samples/bluetooth/peripheral_hids_keyboard/src/main.c

```diff
@@ -0,0 +1,629 @@
+{
+	char addr[BT_ADDR_LE_STR_LEN];
+
+	bt_addr_le_to_str(bt_conn_get_dst(conn), addr, sizeof(addr));
+
+	printk("Security changed: %s level %u\n", addr, level);
+}
+
+
+static struct bt_conn_cb conn_callbacks = {
+	.connected = connected,
+	.disconnected = disconnected,
+	.security_changed = security_changed,
+};
+
+
+
+static void key_pack_queue(char *string, u8_t len)
+{
+	struct key_pack keys;
+	int err;
+
+	__ASSERT(len <= KEYS_MAX_LEN, "String is too long.");
+
+	printk("Getting key: %02X\n", (char) *string);
+
+	keys.elements = string;
+	keys.len = len;
+
+	err = k_msgq_put(&hids_queue, &keys, K_NO_WAIT);
+
+	if (err) {
+		printk("No space in the queue for button pressed\n");
+		return;
+	}
+}
+
+
+static void caps_lock_handler(const struct hids_rep *rep)
+{
+	u8_t report_val = (*rep->data) & OUTPUT_REPORT_BIT_MASK_CAPS_LOCK;
+
+	if (!caps_on && (report_val != 0)) {
+		gpio_pin_write(led_dev, CAPS_LOCK_LED, 0);
+
+		key_pack_queue(caps_on_key_scan_str,
+			       sizeof(caps_on_key_scan_str));
+		caps_on = true;
+	} else if (caps_on && (report_val == 0)) {
+		gpio_pin_write(led_dev, CAPS_LOCK_LED, 1);
+
+		key_pack_queue(caps_off_key_scan_str,
+			       sizeof(caps_off_key_scan_str));
+		caps_on = false;
+	}
+}
+
+
+static void hids_outp_rep_handler(const struct hids_rep *rep,
+				  struct bt_conn *conn)
+{
+	char addr[BT_ADDR_LE_STR_LEN];
+
+	bt_addr_le_to_str(bt_conn_get_dst(conn), addr, sizeof(addr));
+	printk("Output report has been received %s\n", addr);
+	caps_lock_handler(rep);
+}
+
+
+static void hids_boot_kb_outp_rep_handler(const struct hids_rep *rep,
+					  struct bt_conn *conn)
+{
+	char addr[BT_ADDR_LE_STR_LEN];
+
+	bt_addr_le_to_str(bt_conn_get_dst(conn), addr, sizeof(addr));
+	printk("Boot Keyboard Output report has been received %s\n", addr);
+	caps_lock_handler(rep);
+}
+
+
+static void hids_pm_evt_handler(enum hids_pm_evt evt, struct bt_conn *conn)
+{
+	char addr[BT_ADDR_LE_STR_LEN];
+	size_t i;
+
+	for (i = 0; i < CONFIG_BT_GATT_HIDS_MAX_CLIENT_COUNT; i++) {
+		if (conn_mode[i].conn == conn) {
+			break;
+		}
+	}
+
+	if (i >= CONFIG_BT_GATT_HIDS_MAX_CLIENT_COUNT) {
+		return;
+	}
+
+	bt_addr_le_to_str(bt_conn_get_dst(conn), addr, sizeof(addr));
+
+	switch (evt) {
+	case HIDS_PM_EVT_BOOT_MODE_ENTERED:
+		printk("Boot mode entered %s\n", addr);
+		conn_mode[i].in_boot_mode = true;
+		break;
+
+	case HIDS_PM_EVT_REPORT_MODE_ENTERED:
+		printk("Report mode entered %s\n", addr);
+		conn_mode[i].in_boot_mode = false;
+		break;
+
+	default:
+		break;
+	}
+}
+
+
+static void hid_init(void)
+{
+	int err;
+	struct hids_init_param    hids_init_obj = { 0 };
+	struct hids_inp_rep       *hids_inp_rep;
+	struct hids_outp_feat_rep *hids_outp_rep;
+
+	static const u8_t report_map[] = {
+		0x05, 0x01,       /* Usage Page (Generic Desktop) */
+		0x09, 0x06,       /* Usage (Keyboard) */
+		0xA1, 0x01,       /* Collection (Application) */
+		0x05, 0x07,       /* Usage Page (Key Codes) */
+		0x19, 0xe0,       /* Usage Minimum (224) */
```

Comment:
Adding a report id would make sense, as this would be a requirement if you add another class, like consumer control (play/pause/etc), or do a combo mouse+keyboard.

Reply:
Well... the idea with Report ID is simple and defined in HID specification - If you have more than 1 report you are adding report ID. If you have only 1 report, you do not have to do it. It is not a matter in bluetooth, but in USB you even cannot add Report ID for the data stream when you have only 1 report.

Moreover - there is report ID implemented. See lines 240, 268, 286, The code 0x85.

Reply:
In keyboard we are using ID0 that is invalid ID, and we are not storing it in the report map. For me it seems ok. It might be a matter of discussion, but this is just simple, first implementation. And this is work done by me only to test HID central that is my real task. I prefer to rather vote to create another task for such a work.

Reply:
ReportID 0 is an indication that the reportid is not set at all, and is only accepted when its one report in the descriptor and no explicit ReportID defined. If you define it, it must be 1 or larger. Explicitly setting reportid to '1' is a small improvement that makes the example easier to build upon, especially for developers that are not very experienced with usb.

Reply:
No - it is not an improvement - it just hides some issues that inexperienced user may experience.
Anyway - to cut the discussion here - I have implemented option to define the ID and put it into the Report Map when used.